convert2rhel/subscription.py

```diff
@@ -278,6 +287,93 @@ def args(self):
+        # * If we fallback to cli, do we still allow activation_key even though it is insecure?
+        # * Should checking for DBUS be done in either system_info or checks.py?
+        # * What should the return value of this code be?  Should cli non-zero retcodes raise an
+        #   exception instead of returning the code?
+        # TODO:
+        # * In _call_via_dbus(), we need to add more logging.  Call_via_cli lets
+        #   run_subprocess()/run_in_pty() do most of its logging but we don't have that in
+        #   _call_via_dbus()
+        if DBUS:
+            return self._call_via_dbus()
+        else:
+            return self._call_via_cli()
+
+    def _call_via_dbus(self):
+        """
+        Use dbus to register the system with subscription-manager.
+
+        Status of dbus on various platforms:
+            * RHEL6:
+                * DBUS is present but may not be installed.
+                * Install the dbus package and run /etc/rc.d/init.d/messagebus start
+                * dbus-python-0.83.0 is available
+            * RHEL7:
+                * dbus looks to be installed and run by default
+                * dbus-python-1.1.9 is available
+            * RHEL8:
+            * RHEL9:
+                * dbus looks to be installed and run by default
+                * python3-dbus-1.2.18 is available
+
+        .. seealso::
+            Documentation for the subscription-manager dbus API:
+            https://www.candlepinproject.org/docs/subscription-manager/dbus_objects.html#registerserver
+        """
+        system_bus = dbus.SystemBus()
+
+        if self.server_url:
+            connection_opts = {"host": self.server_url}
+        else:
+            # DBus seems to want empty string if there are no values even though the type is a dict
+            connection_opts = ""
+
+        # Create a new bus so we can talk to rhsm privately (For security:
+        # talking on the system bus might be eavesdropped in certain scenarios)
+        register_server = system_bus.get_object("com.redhat.RHSM1", "/com/redhat/RHSM1/RegisterServer")
+        address = register_server.Start("", dbus_interface="com.redhat.RHSM1.RegisterServer")
+
+        try:
+            # Use the private bus to register the machine
+            private_bus = dbus.connection.Connection(address)
+            register_object = private_bus.get_object("com.redhat.RHSM1", "/com/redhat/RHSM1/Register")
+
+            if self.password:
+                register_object.Register(
+                    self.organization or "",
+                    self.username,
+                    self.password,
+                    "",
+                    connection_opts,
+                    "C",
+                    dbus_interface="com.redhat.RHSM1.Register",
+                )
+
+            else:
+                register_object.RegisterWithActivationKeys(
+                    self.org,
+                    [self.activation_key],
+                    "",
```

Comment:
`options` is a dict, so this must be an empty dict rather than an empty string

Reply:
Pushed a change that changes this to dbus.Dictionary({}, signature="ss")